src/result.ts

```diff
@@ -314,10 +377,26 @@ export function Ok<T, E extends Error>(input?: Exclude<T, E>) {
  * }
  *
  * ```
+ *
+ * @example
+ * ```ts
+ * const foo = Err(new Error("Foo!"));
+ *
+ * if (foo instanceof Err) {
+ *  // Do something
+ * }
+ * ```
```

Comment:
This function could produce an Ok.

```TypeScript
const a = 2 as unknown as Error
const res = Err(a) // Not a string, not an Error, but TypeScript thinks so
console.log(res.isOk() // true
console.log(res.isErr() // false
```

Perhaps a generic `isErrorLike(input: any): input is Error` function could be created which checks for characteristics of objects like Errors could be used in this function as well as the `Result.isErr` function.

```TypeScript
interface ErrorKind {
	message: string;
}
type ErrorLike = Error | ErrorKind;

function isErrorLike(input: any): input is ErrorLike {
	return (
		input instanceof Error ||
		(input &&
			typeof input === 'object' &&
			(Object.prototype.isPrototypeOf.call(Error, input) ||
				typeof input.message === 'string'))
	);
}
```
I am kind of concerned about forcing Error as the only type here, as then we get undefined behavior at runtime anyway... should it throw if anything not Error or string is passed?

I think Rust allows any value to be passed for E in Result<T, E>, and JS allows us to throw any value, so perhaps there could be some other signal that what's constructed is an error?

Perhaps an optional `type: 'Err' | 'Ok' = 'Ok'` as an arg, and then always pass `'Err'` from `function Err`?

Such a type value could then also be used in the `Result.isOk` and `Result.isErr` functions to avoid prototype chain lookups which are currently used to decide `Result.isErr`. Might provide a performance bump, and it's how I do it in my private lib.

Reply:
> ```ts
>const a = 2 as unknown as Error
>const res = Err(a) // Not a string, not an Error, but TypeScript thinks so
>console.log(res.isOk() // true
>console.log(res.isErr() // false
>```

I don't think we should worry about someone performing incorrect type casting - `as` is something that TS developers should try to avoid.

> Perhaps a generic isErrorLike(input: any): input is Error function could be created which checks for characteristics of objects like Errors could be used in this function as well as the Result.isErr function.
>```ts
>interface ErrorKind {
>	message: string;
>}
>
>type ErrorLike = Error | ErrorKind;
>
>function isErrorLike(input: any): input is ErrorLike {
>	return (
>		input instanceof Error ||
>		(input &&
>			typeof input === 'object' &&
>			(Object.prototype.isPrototypeOf.call(Error, input) ||
>				typeof input.message === 'string'))
>	);
>}
>```
>I am kind of concerned about forcing Error as the only type here, as then we get undefined behavior at runtime anyway... should it throw if anything not Error or string is passed?
>
>I think Rust allows any value to be passed for E in Result<T, E>, and JS allows us to throw any value, so perhaps there could be some other signal that what's constructed is an error?
>
>Perhaps an optional type: 'Err' | 'Ok' = 'Ok' as an arg, and then always pass 'Err' from function Err?
>
>Such a type value could then also be used in the Result.isOk and Result.isErr functions to avoid prototype chain >lookups which are currently used to decide Result.isErr. Might provide a performance bump, and it's how I do it in >my private lib.

This is very interesting and should definitely be addressed in a separate issue! My thoughts at the time of sticking to `Error` is that its the native JS way of error handling and it provides the unique benefit of a call stack. 

You are definitely correct that other values should be returnable in Err, the question is how restrictive should it be? I think we should 100% at the very least allow `enum`s to be passed.